runtime/error.c

```diff
@@ -1,9 +1,171 @@
+            functionFound = true;
+        }
+        if (filenameFound && functionFound)
+            return;
+    }
+}
+
+static void onError(void *data, const char *msg, int errnum) {
+    fprintf(stderr, "Error : libbacktrace : %s\n", msg);
+    fflush(stderr);
+}
+
+static int onFrame(void *data, uintptr_t pc, const char *filename, int lineno, const char *function) {
+    // This check is valid if there is a posibility not to have debug info
+    if (filename == NULL || function == NULL) {
+        return 0;
+    }
+    Trace *trace = (Trace *)data;
+    uint32_t nFrames = trace->nFrames;
+    uint32_t szFrames = trace->szFrames;
+    if (unlikely(nFrames == MAX_FRAME_COUNT))
+        return 1;
+    if (nFrames == szFrames) {
+        szFrames = szFrames << 1;
+        trace->szFrames = szFrames;
+        void *p = realloc(trace->frames, sizeof(Frame) * szFrames);
+        if (p == 0)
+            return 1;
+        trace->frames = (Frame *)p;
+    }
+
+    uint32_t nBytes = trace->nBytes;
+    uint32_t szBytes = trace->szBytes;
+    uint64_t filenameLen = strlen(filename);
+    uint64_t funcnameLen = strlen(function);
+    uint32_t requiredBytes = nBytes + filenameLen + funcnameLen + 2;
+    if (unlikely(requiredBytes > MAX_BYTES_COUNT))
+        return 1;
+    if (requiredBytes > szBytes) {
+        szBytes = szBytes << 1;
+        while (requiredBytes > szBytes) {
+            szBytes = szBytes << 1;
+        }
+        trace->szBytes = szBytes;
+        void *p = realloc(trace->bytes, szBytes);   
+        if (p == 0)
+            return 1;
+        trace->bytes = (char *)p;
+    }
+
+    Frame *frame = trace->frames + nFrames;
+    frame->pc = pc;
+    frame->lineno = lineno;
+
+    Offset offset = {-1, -1};
+    char *bytes = trace->bytes;
+    checkForSameStr(frame, nFrames, filename, function, bytes, &offset);
+
+    if (offset.fileOffset >= 0) {
+        frame->fileOffset = offset.fileOffset;
+    }
+    else {
+        frame->fileOffset = nBytes;
+        memcpy(bytes + nBytes, filename, filenameLen);
+        nBytes = nBytes + filenameLen;
+        *(bytes + nBytes) = '\0';
+        nBytes++;
+    }
+
+    if (offset.functionOffset >= 0) {
+        frame->functionOffset = offset.functionOffset;
+    } 
+    else {
+        frame->functionOffset = nBytes;
+        memcpy(bytes + nBytes, function, funcnameLen);
+        nBytes = nBytes + funcnameLen;
+        *(bytes + nBytes) = '\0';
+        nBytes++;
+    }
+
+    trace->nBytes = nBytes;
+    trace->nFrames = nFrames + 1;
+
+    return 0;
+}
+
+static int getBackTrace(Trace* trace) {
+    state = backtrace_create_state(NULL, THREAD, onError, NULL);
+
+    void *p = malloc(sizeof(Frame) * INITIAL_FRAME_COUNT);
+    if (p == 0)
+        return 1;
+    trace->frames = (Frame *)p;
+    p = malloc(INITIAL_BYTES_COUNT);
+    if (p == 0)
+        return 1;
+    trace->bytes = (char *)p;
+
+    return backtrace_full(state, SKIP_FROM_END, onFrame, onError, trace);
```

Comment:
In the last call, we concluded that we should call backtrace_simple when  constructing an error so as to store just the pc. Then we would use backtrace_pcinfo or backtrace_syminfo (according to whether we have debugging info or no), which is the expensive part, when we need to print it out.

Reply:
@jclark , in the `_bal_error_constructor` we only have to construct the error with pcs and we have to create a new method like `_bal_error_trace_print` in error.c. Is that correct?

Reply:
Right.

Reply:
`<function>(<filepath>:<lineno>)`
@jclark, Is above format looks okay to when printing backtrace?

Reply:
You shouldn't need to allocate while printing: just print it out from the callback.

Reply:
Looks ok. Have a look at what some other languages do.

Note that you should not need to allocate when printing out: just print directly from the backtrace.  Eventually we will need to demangle the function name, but demangling will need to change to support multiple modules, so let's leave this till later.

Reply:
Something simple is enough for now.

In the long term we'll need to design this properly. On the jBallerina side we have done some changes recently to include cause in the trace, and also to remove duplicate entities when we are doing so (similar to Java). There were some requests from debugger fork asking if we can show entries from parent strand, which we haven't done, but maybe that is a useful thing? Feel like we need a place to document this, is spec the correct place?

Reply:
> Looks ok. Have a look at what some other languages do.
> 
Error trace for java and rust is shown below.
Java:
```
<ERROR_TYPE>
        at <function>(<filename>:<lineno>)
```

Rust:
```
<#number> <function>
       at <filepath>:<lineno>:<columnno>

```
Go has a similar format as shown above. In python, function name appears at the end of the trace line, but lineno always follow the filepath.

Reply:
I agree lineno should follow filepath.

I don't care for putting the filename/lineno in parentheses. Implement whatever you prefer and we can adjust later.